more info on hardware requirements (#9145)
opf/openproject: docs/installation-and-operations/system-requirements/README.md

## Server

The server hardware requirements should be rougly the same for both the packaged installation and docker (both all-in-one container and compose).

### Minimum Hardware Requirements

* __CPU:__ Quad Core CPU (>= 2ghz)
* __Memory:__ 4096 MB
* __Free disk space:__ 20 GB

> This is for a single Server running OpenProject for up to 200 total active users. Depending on your number of peak users, these requirements might vary drastically.

Reviewer: Typo: Server -> server

> This is for a single Server running OpenProject for up to 200 total active users. Depending on your number of peak users, these requirements might vary drastically.

Reviewer 2: ```suggestion
This is for a single server running OpenProject for up to 200 total users. Depending on your number of concurrent users,  these requirements might vary drastically.
```

I reworded this sentence since I find `peak` not to be really clear. After some googling, performance statistics often seem to mention either _concurrent users_ (performing any request on the page within the timeframe) and _simultaneous users_ (number of users performing the same request at the same time). I'd say we stick with `concurrent users` then.

### General Requirements

Generally speaking you will need more CPUs (the faster the better) and more RAM with an increasing number of users.

> Technically this only really depends on the number of __peak__ users. No matter if you have 1000 or only 100 users overall, if there only ever 20 users working at the same time, the CPU & RAM required will be the same.

Reviewer 2: ```suggestion
Technically this only really depends on the number of concurrent users. No matter if you have 1000 or only 100 total users, if there only ever 20 users working at the same time, the CPU & RAM required will be the same.
```

> Still, the total number of users is a good general indicator of how much resources you will need.

Reviewer: I'd add a section that it is not simply a "one worker per concurrent user" equation as almost all pages require multiple requests. So even for one only one concurrent user, I'd recommend having more than one worker available because of that.

It's not enough to simply have more resources available, however. You will have to make use of them too.

> By default OpenProject has 4 so called web workers and 1 background worker.

Reviewer 2: ```suggestion
By default OpenProject has 4 so called web workers and 1 background worker. Web workers are handling the HTTP requests while backend workers are doing offloaded tasks such as sending emails or performing resource-intensive tasks of unknown duration, e.g., copying or deleting resources.
```

If there are more users you will need more web workers and eventually also more background workers.

The database will need resources as well, and this, too, will increase with the number of users.

> There may come a point where you will have to make configuration changes to the database and/or use an external database, but for most cases the default database setup should be enough.

Reviewer 2: ```suggestion
There may come a point where you will have to make configuration changes to the database and/or use an external database, but for most cases the default database setup should be enough. You will ideally want to have the database on a performant storage such as SSDs. [There are also other excellent resources](https://wiki.postgresql.org/wiki/Performance_Optimization) for tuning PostgreSQL database performance.
```

Using a rough estimate we can give the following recommendations based on the number of users.

| Users | CPU cores | RAM in GB  | web workers | background workers | Diskspace in GB |
|-------|-----------|------------|-------------|--------------------|-----------------|
| 200   | 4         | 4          | 4           | 1                  | 20              |
| 500   | 8         | 8          | 8           | 2                  | 40              |

> | 1500  | 16        | 16         | 16          | 4                  | 80              |

Reviewer: Because one user request can lead to multiple browser requests I'd increase the number of web workers at least for the 200 users scenario. 
I think I'd start with 4 workers for up to 20 (or maybe even 50) users and go to 8 for 200. 16 for 500 and 32 for 1500. The factor of requiring multiple actual requests per user triggered request will probably balance out better the more users are involved. 
It is probably better to have higher requirements then to later have to deal with performance complaints.

Reviewer 2: Previously we've been talking about total or concurrent users. Is this table targeted at total or concurrent users? For concurrent, the values would be much higher and I'd stop at 500 / 32gig

Author: The table is aimed at total users. I should perhaps amend the heading in that case. Concurrent users is nice in theory practically people will find that hard to answer. At least hard than total users.

> Using a rough estimate we can give the following recommendations based on the number of users.
> 
> | Users | CPU cores | RAM in GB  | web workers | background workers | Diskspace in GB |
> |-------|-----------|------------|-------------|--------------------|-----------------|
> | 200   | 4         | 4          | 4           | 1                  | 20              |
> | 500   | 8         | 8          | 8           | 2                  | 40              |
> | 1500  | 16        | 16         | 16          | 4                  | 80              |

Reviewer 2: ```suggestion
Using a rough estimate we can give the following recommendations based on the number of concurrent users.

| Concurrent users | CPU cores | RAM in GB  | web workers | background workers | Diskspace in GB |
|-------|-----------|------------|-------------|--------------------|-----------------|
| 20   | 4         | 4          | 4           | 1                  | 20              |
| 50   | 8         | 8          | 8           | 2                  | 40              |
| 100  | 16        | 16         | 16          | 4                  | 80              |
| 200  | 32        | 32         | 32          | 8                  | 150              |
```

Suggestion for concurrent users

These numbers are a guideline only and your mileage may vary.<sup>1</sup>
It's best to monitor your server and its resource usage. You can always allocate more resources if needed.

See [here](/installation-and-operations/operation/control/#scaling-the-number-of-web-workers) how to scale those up in a packaged installation. If you are using docker-compose you can [scale](https://docs.docker.com/compose/reference/scale/) the web and worker services too.

> <sup>1</sup> When using [docker-compose](https://github.com/opf/openproject-deploy/tree/stable/11/compose) (with `USE_PUMA=true`) you can use fewer web workers which may use a bit more RAM, however. For instance for 200 users a single web worker would be enough.

**Scaling horizontally**

> At some point simply increasing the resources of one single server may not be enough anymoe.

Reviewer: Typo: anymo -> anymore.

## Server

The server hardware requirements should be rougly the same for both the packaged installation and docker (both all-in-one container and compose).

### Minimum Hardware Requirements

* __CPU:__ Quad Core CPU (>= 2ghz)
* __Memory:__ 4096 MB
* __Free disk space:__ 20 GB

> This is for a single Server running OpenProject for up to 200 users.

Reviewer 2: ```suggestion
This is for a single Server running OpenProject for up to 200 total active users. Depending on your number of peak users, these requirements might vary drastically.
```
